File: app_canibuythat/src/main/java/com/gb/canibuythat/provider/BalanceCalculator.java
package com.gb.canibuythat.provider;

import com.gb.canibuythat.model.Spending;
import com.gb.canibuythat.util.DateUtils;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.Date;
import java.util.List;

public class BalanceCalculator {

    /**
     * Calculate how many times the specified <code>spending</code> has been
     * applied up until the specified <code>end</code> date (usually today) and sum up
     * it's value.
     *
     * @param end if null, `today` is used instead
     * @return two floating point values, the first is the minimum possible value, the second is
     * the maximum possible value
     */
    public static BalanceResult getEstimatedBalance(Spending spending, Date start, Date end) {
        if (start != null && end != null && !end.after(start)) {
            throw new IllegalArgumentException("Start date must come before end date!");
        }
        boolean exit = false;
        float bestCase = 0;
        float worstCase = 0;
        List<Date> spendingEvents = new ArrayList<>();
        int count = 0;
        Calendar occurrenceStart = Calendar.getInstance();
        occurrenceStart.setTime(spending.getFromStartDate());
        Calendar occurrenceEnd = Calendar.getInstance();
        occurrenceEnd.setTime(spending.getFromEndDate());
        Calendar date = Calendar.getInstance();
        if (end != null) {
            date.setTime(end);
        }
        DateUtils.clearLowerBits(date);
        start = start != null ? DateUtils.clearLowerBits(start) : null;
        do {
            if (start == null || occurrenceEnd.getTimeInMillis() >= start.getTime()) {
                int r = DateUtils.compare(date, occurrenceStart, occurrenceEnd);
                if (r >= -1) { //after or on start date
                    if (spending.getEnabled()) {
                        worstCase += spending.getValue();
                    }
                    if (r > 1) { // after end date
                        if (spending.getEnabled()) {
                            bestCase += spending.getValue();
                        }
                        spendingEvents.add(occurrenceEnd.getTime());
                    }
                } else {
                    exit = true;
                }
            }
            spending.getCycle().apply(occurrenceStart, spending.getCycleMultiplier());
            spending.getCycle().apply(occurrenceEnd, spending.getCycleMultiplier());
            if (spending.getOccurrenceCount() != null && ++count >= spending.getOccurrenceCount()) {
                exit = true;
            }
        } while (!exit);
        return new BalanceResult(bestCase, worstCase, spendingEvents.toArray(new Date[spendingEvents.size()]));
    }

    public static class BalanceResult {
        /**
         * Minimum amount of money that is expected to be spent until now (according to
         * the user's spending habits) on a particular spending category. The
         * difference between minimum and maximum comes from the fact that a
         * SpendingItem may be spent in an interval rather than on an exact date.
         */
        public final float bestCase;
        /**
         * Maximum amount of money that is expected to be spent until now (according to
         * the user's spending habits) on a particular spending category. The
         * difference between minimum and maximum comes from the fact that a
         * SpendingItem may be spent in an interval rather than on an exact date.<br>
         */
        public final float worstCase;
        /**
         * Spending event is the latest day on which we expect a payment/income to happen.
         * For eg if a weekly Spending starts on a Monday, than the array of spending
         * events will be all the Sundays after that Monday and before today (including
         * today).
         */
        public final Date[] spendingEvents;

        BalanceResult(float bestCase, float worstCase, Date[] spendingEvents) {
            this.bestCase = bestCase;
            this.worstCase = worstCase;
            this.spendingEvents = spendingEvents;
        }
    }
}

File: app_canibuythat/src/main/java/com/gb/canibuythat/provider/BalanceCalculator.kt
package com.gb.canibuythat.provider

import com.gb.canibuythat.model.Spending
import com.gb.canibuythat.util.DateUtils
import com.gb.canibuythat.util.clearLowerBits
import com.gb.canibuythat.util.toCalendar
import java.util.*

object BalanceCalculator {

    /**
     * Calculate how many times the specified `spending` has been
     * applied up until the specified `end` date (usually today) and sum up
     * it's value.

     * @param end if null, `today` is used instead
     * *
     * @return two floating point values, the first is the minimum possible value, the second is
     * * the maximum possible value
     */
    fun getEstimatedBalance(spending: Spending, startDate: Date?, endDate: Date?): BalanceResult {
        if (startDate != null && endDate != null && !endDate.after(startDate)) {
            throw IllegalArgumentException("Start date must come before end date!")
        }

        var `break` = false
        var definitelySpentThisMuch = 0f
        var maybeSpentThisMuch = 0f
        val spendingEvents = ArrayList<Date>()
        var count = 0
        val occurrenceStart = spending.fromStartDate.toCalendar()
        val occurrenceEnd = spending.fromEndDate.toCalendar()
        val start = startDate?.clearLowerBits() ?: null
        val end: Calendar = endDate?.toCalendar() ?: Calendar.getInstance()
        end.clearLowerBits()
        do {
            if (start == null || occurrenceEnd.timeInMillis >= start.time) {
                val r = DateUtils.compare(end, occurrenceStart, occurrenceEnd)
                if (r >= -1) { // >= start date
                    if (spending.enabled) {
                        maybeSpentThisMuch += spending.value.toFloat()
                    }
                    if (r > 1) { // > end date
                        if (spending.enabled) {
                            definitelySpentThisMuch += spending.value.toFloat()
                        }
                        spendingEvents.add(occurrenceEnd.time)
                    }
                } else {
                    `break` = true
                }
            }
            spending.cycle.apply(occurrenceStart, spending.cycleMultiplier!!)
            spending.cycle.apply(occurrenceEnd, spending.cycleMultiplier!!)
            if (spending.occurrenceCount != null && ++count >= spending.occurrenceCount!!) {
                `break` = true
            }
        } while (!`break`)
        return BalanceResult(definitelySpentThisMuch, maybeSpentThisMuch, spendingEvents.toTypedArray())
    }

    data class BalanceResult(
            /**
             * We know for certain that this amount has been spent by the selected date
             */
            val definitely: Float,
            /**
             * `maybeEvenThisMuch` is `definitely` plus spendings that may have been spent by the selected date.
             * A weekly spending for eg. "may be spent" for one week after which it becomes "definitely spent"
             * and is included only with the `definitely` sum.
             */
            val maybeEvenThisMuch: Float,
            /**
             * Spending event is the latest day on which we expect a payment/income to happen.
             * For eg if a weekly Spending starts on a Monday, than the array of spending
             * events will be all the Sundays after that Monday and before today (including
             * today).
             */
            val spendingEvents: Array<Date>)
}
